### ISSUE TITLE
Lens support

### ISSUE THREAD

**Issue by mf59816** [Soostone/uri-bytestring#2]
I can (probably) write some code, but I wanted to ask your opinion first.  I see two options:

1. depend on lens and use it like in e.g. snap
2. provide extra package uri-bytestring-lens on top of uri-bytestring

Would you accept a PR for 1.?  Did I overlook somebody having solved this already?

thanks,
matthias

**Comment by ozataman** [comment]
Any chance we can add lenses without depending on lens explicitly?

**Comment by mf59816** [comment]
good idea.  i think what we need is essentially a bunch of functions like this:

    uriPath :: Lens' URI Path
    uriQuery :: Lens' URI Query

the types are not particularly magic:

    type Lens s t a b = forall f. Functor f => (a -> f b) -> s -> f t
    type Lens' s a = Lens s s a a

and call the record field selectors `_uriPath` etc.  can you see the implementations?  I'm a bit too rushed and stupid right now, but I could give it a shot later.

(if i do: should this go into a separate module?)

**Comment by mightybyte** [comment]
@mf59816 If you write this, I would recommend using the style I use [here](https://github.com/snapframework/heist/blob/master/src/Heist/Internal/Types.hs) which would allow us to supply lenses without depending on a specific lens library.

**Comment by MichaelXavier** [comment]
@mf59816 We discussed this a bit and we would want a URI.ByteString.Lens module with generic lenses. It would also be really nice to have a `Prism URI ByteString` that uses the parser/serializer provided in the main module. Lens has a great writeup of all the generic types for lens here:

https://github.com/ekmett/lens/wiki/How-can-I-write-lenses-without-depending-on-lens%3F

I may take a crack at this today if I have the time. I'll tag this ticket with the fix for everyone's review.

**Comment by MichaelXavier** [comment]
Feel free to pull down the lens branch and try this out, look at the haddocks and see how it feels. I've added Lenses, Isos, and Prisms where appropriate. If everything looks good, we can merge this and cut a release (probably 0.1)

**Comment by mightybyte** [comment]
Looks pretty good at first glance.  The profunctors dependency needed for isos and prisms is a bit annoying.  I think I'm leaning towards the opinion that for this library it's too much and we should just stick to the lenses.  But I'm open to being convinced otherwise.

**Comment by MichaelXavier** [comment]
I guess its a matter of what we are worried about. The dependency chain for profunctors is full of  extremely common packages (transformers, containers, array) and ones with few to no dependencies (tagged, nats). So it seems to me like the dependency hell risk and compile time impact isn't too bad. There may be some other reason for concern that I'm not considering. If it is significant though, I advocate moving this stuff to uri-bytestring-lens or some such.

**Comment by mightybyte** [comment]
I think people are going to see that profunctors has almost 60% of the dependencies that lens has and immediately lump this into the same category.  In exchange for this cost, we only get 4 simple isos and one prism.  To me that doesn't seem worth it.  We want this to be the de facto standard uri library, so I think minimizing dependencies is more important than a handful of one-line convenience functions.  I don't think we need to move it to a separate package either because we can still keep most of the value here without paying that cost.

**Comment by mf59816** [comment]
Cool, thanks!

I tend to agree with mightybyte on the question of depending on profunctor, but it wouldn't bother me personally.

Why does `queryPairs` has a new name, and other lenses use the old name with trailing underscore?  Is there a reason you broke symmetry that I am missing?

The underscore oftten indicates a variant of a monadic function that substitutes the return type with m ().  I would prefer `lgetQuery` or `lGetQuery` or `getQueryL` (even though that may be less idiomatic).

But those are minor points.  I would be perfectly happy with a release of what you have.

**Comment by mightybyte** [comment]
Ahh yes.  I agree with @mf59816 about the trailing underscores.  For lenses that correspond exactly to our field names I think we should adopt the lens convention and put a leading underscore on the field names and drop it for the lenses.  The library has only been out for a month and downloaded just 36 times so far.  Our conventions should be as mainstream as possible and now is the time when breaking things is still cheap.

**Comment by MichaelXavier** [comment]
I've updated the branch with the following changes:

1. Dropped profunctors.
2. Eliminated prisms.
3. Exchanged newtype Isos for Lenses
4. Added inlining for each lens function under Ed's recommendation.
5. Renamed all the direct fields to start with `_`, pulling the bandage off now rather than later.
6. Renamed some of the iso field names to be more sensible with lenses. i.e. hostBS instead of getHost because it would be weird to do a lens update with a getter and also weird to have the lens disagree with the field name.

If everything looks good let me know and I'll merge.
